plugin/pkg/auth/authorizer/webhook/webhook.go

```diff
@@ -0,0 +1,141 @@
+limitations under the License.
+*/
+
+// Package webhook implements the authorizer.Authorizer interface using HTTP webhooks.
+package webhook
+
+import (
+	"bytes"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/golang/glog"
+
+	"k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/apis/authorization/v1beta1"
+	"k8s.io/kubernetes/pkg/auth/authorizer"
+	"k8s.io/kubernetes/pkg/util"
+)
+
+// Ensure Webhook implements the authorizer.Authorizer interface.
+var _ authorizer.Authorizer = (*WebhookAuthorizer)(nil)
+
+type WebhookAuthorizer struct {
+	callback string
+	cli      *http.Client
+}
+
+func New(callback, clientCertFile, clientKeyFile, caFile string) (*WebhookAuthorizer, error) {
+	var clientTLSConfig tls.Config
+
+	if clientCertFile != "" && clientKeyFile != "" {
+		clientCert, err := tls.LoadX509KeyPair(clientCertFile, clientKeyFile)
+		if err != nil {
+			return nil, err
+		}
+		clientTLSConfig.Certificates = []tls.Certificate{clientCert}
+	} else {
+		glog.Info("No x509 certificate provided for client identity.")
+	}
+
+	if caFile != "" {
+		roots, err := util.CertPoolFromFile(caFile)
+		if err != nil {
+			return nil, err
+		}
+		clientTLSConfig.RootCAs = roots
+	} else {
+		glog.Info("No x509 certificates provided, will use host's root CA set")
+	}
+
+	// TODO(ericchiang): Can we check here to ensure remote service is reachable?
+
+	return &WebhookAuthorizer{
+		callback: callback,
+		cli: &http.Client{
+			// Same as http.DefaultTransport except for TLSClientConfig.
+			Transport: &http.Transport{
+				Proxy: http.ProxyFromEnvironment,
+				Dial: (&net.Dialer{
+					Timeout:   30 * time.Second,
+					KeepAlive: 30 * time.Second,
+				}).Dial,
+				TLSHandshakeTimeout: 10 * time.Second,
+				TLSClientConfig:     &clientTLSConfig,
+			},
+		},
+	}, nil
+}
+
+// RequestError indicates an error composing, sending, or decoding the response of a request.
+type RequestError struct {
+	msg string
+}
+
+func (r *RequestError) Error() string {
+	return r.msg
+}
+
+func logErrorf(format string, a ...interface{}) *RequestError {
+	msg := fmt.Sprintf(format, a...)
+	glog.Error(msg)
+	return &RequestError{msg}
+}
+
+func (w *WebhookAuthorizer) Authorize(attr authorizer.Attributes) (err error) {
+	r := v1beta1.SubjectAccessReviewSpec{
+		User:   attr.GetUserName(),
+		Groups: attr.GetGroups(),
+	}
+	if attr.IsResourceRequest() {
+		r.ResourceAttributes = &v1beta1.ResourceAttributes{
+			Namespace: attr.GetNamespace(),
+			Verb:      attr.GetVerb(),
+			Group:     attr.GetAPIGroup(),
+			Resource:  attr.GetResource(),
+		}
+	} else {
+		r.NonResourceAttributes = &v1beta1.NonResourceAttributes{
+			Path: attr.GetPath(),
+			Verb: attr.GetVerb(),
+		}
+	}
+
+	data, err := json.Marshal(r)
+	if err != nil {
+		return logErrorf("Failed to serizalize authz webhook request: %v", err)
+	}
+	resp, err := w.cli.Post(w.callback, "application/json", bytes.NewReader(data))
```

Comment:
Thanks for weighing in. Yeah, I forgot to mention this.

On Tue, Feb 2, 2016 at 5:36 PM, Jordan Liggitt <notifications@github.com>
wrote:

> In plugin/pkg/auth/authorizer/webhook/webhook.go
> <https://github.com/kubernetes/kubernetes/pull/20347#discussion_r51667097>
> :
>
> > +			Verb:      attr.GetVerb(),
> > +			Group:     attr.GetAPIGroup(),
> > +			Resource:  attr.GetResource(),
> > +		}
> > +	} else {
> > +		r.NonResourceAttributes = &v1beta1.NonResourceAttributes{
> > +			Path: attr.GetPath(),
> > +			Verb: attr.GetVerb(),
> > +		}
> > +	}
> > +
> > +	data, err := json.Marshal(r)
> > +	if err != nil {
> > +		return logErrorf("Failed to serizalize authz webhook request: %v", err)
> > +	}
> > +	resp, err := w.cli.Post(w.callback, "application/json", bytes.NewReader(data))
>
> If we're going to implement a remote authorizer, especially using these
> types, we should use the types as intended, and expect a
> SubjectAccessReview response, and use the spec.allowed as the answer.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20347/files#r51667097>.
>